csi/controller.go

```diff
@@ -980,6 +1016,88 @@ func (s *OsdCsiServer) CreateSnapshot(
 		},
 	}, nil
 }
+func (s *OsdCsiServer) getCloudBackupClient(ctx context.Context) (api.OpenStorageCloudBackupClient, error) {
+	// Get grpc connection
+	conn, err := s.getRemoteConn(ctx)
+	if err != nil {
+		return nil, status.Errorf(
+			codes.Unavailable,
+			"Unable to connect to SDK server: %v", err)
+	}
+	return s.cloudBackupClient(conn), nil
+}
+
+func (s *OsdCsiServer) createCloudBackup(
+	ctx context.Context,
+	req *csi.CreateSnapshotRequest,
+) (*csi.CreateSnapshotResponse, error) {
+	cloudBackupClient, err := s.getCloudBackupClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get any labels passed in by the CO
+	_, locator, _, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "Unable to get parameters: %v", err)
+	}
+
+	backupID := req.GetName()
+	// Create snapshot
+	_, err = cloudBackupClient.Create(ctx, &api.SdkCloudBackupCreateRequest{
+		VolumeId: req.GetSourceVolumeId(),
+		TaskId:   backupID,
+		Labels:   locator.GetVolumeLabels(),
+	})
+
+	if err != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to create cloud snapshot: %v", err)
+	}
+
+	var isBackupReady bool
+	var backupStatus *api.SdkCloudBackupStatusResponse
+
+	// Check if snapshot has been created but is in error state
+	backupStatus, errFindFailed := cloudBackupClient.Status(ctx, &api.SdkCloudBackupStatusRequest{
+		VolumeId: req.GetSourceVolumeId(),
+		TaskId:   backupID,
+	})
+	if errFindFailed != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to create cloud snapshot: %v", err)
+	}
+	isBackupReady = backupStatus.Statuses[backupID].Status == api.SdkCloudBackupStatusType_SdkCloudBackupStatusTypeDone ||
+		backupStatus.Statuses[backupID].Status == api.SdkCloudBackupStatusType_SdkCloudBackupStatusTypeActive
+	if errFindFailed == nil && !isBackupReady {
+
+		logrus.Infof("cleaned up failed cloud snapshot %v in order to retry snapshot create", backupID)
+		// If snapshot was created but has errors, cleanup and re-create it on the next iteration.
+		_, errCleanupSnap := cloudBackupClient.Delete(ctx, &api.SdkCloudBackupDeleteRequest{
+			BackupId: backupID,
+		})
+		if errCleanupSnap != nil {
+			return nil, status.Errorf(codes.Aborted, "Snapshot create failed and unable to cleanup failed cloud snapshot %s: %v",
+				backupID,
+				errCleanupSnap)
+		}
+	}
+
+	snapSize, errSizeFailed := cloudBackupClient.Size(ctx, &api.SdkCloudBackupSizeRequest{
+		BackupId: backupID,
+	})
+	if errSizeFailed != nil {
+		return nil, status.Errorf(codes.Aborted, "Failed to get cloud snapshot size: %v", err)
+	}
+
+	return &csi.CreateSnapshotResponse{
+		Snapshot: &csi.Snapshot{
+			SizeBytes:      int64(snapSize.GetSize()),
```

Comment:
You are using a deprecated size field from the response
```
message SdkCloudBackupSizeResponse{
  // Same value as total_download_bytes, will be deprecated
  uint64 size = 1;
  // Total download size of all the backups (full & incremental) required for a restore
  uint64 total_download_bytes = 2;
  // Size of the compressed individual backup in object-store or NFS
  uint64 compressed_object_bytes = 3;
  // Used size of the volume to be restored (in bytes)
  uint64 capacity_required_for_restore = 4;
}
```
According to CSI spec, what is the expected value from the size field, since it needs to map to one of these values.

Reply:
Are you planning to fix this?